File: kotlin-archive/jam-pvm/src/main/kotlin/io/forge/jam/pvm/program/LookupTable.kt
package io.forge.jam.pvm.program

/**
 * Type alias for lookup entry represented as an unsigned 32-bit integer
 */
typealias LookupEntry = UInt

/**
 * Constant representing an empty lookup entry
 */
private const val EMPTY_LOOKUP_ENTRY: LookupEntry = 0u

/**
 * A lookup table for handling immediate value operations.
 * This class provides efficient lookup capabilities for immediate value bit patterns.
 *
 * @property table The underlying array of lookup entries
 */
@JvmInline
value class LookupTable private constructor(
    private val table: Array<LookupEntry>
) {
    companion object {
        /**
         * Packs immediate value parameters into a single lookup entry.
         *
         * @param imm1Bits First immediate value bits
         * @param imm1Skip First immediate value skip amount
         * @param imm2Bits Second immediate value bits
         * @return Packed lookup entry
         */
        private fun pack(imm1Bits: UInt, imm1Skip: UInt, imm2Bits: UInt): LookupEntry {
            require(imm1Bits <= 0b111111u) { "imm1Bits out of range" }
            require(imm2Bits <= 0b111111u) { "imm2Bits out of range" }
            require(imm1Skip <= 0b111111u) { "imm1Skip out of range" }

            return imm1Bits or (imm1Skip shl 6) or (imm2Bits shl 12)
        }

        /**
         * Unpacks a lookup entry into its component values.
         *
         * @param entry The lookup entry to unpack
         * @return Triple of (imm1Bits, imm1Skip, imm2Bits)
         */
        private fun unpack(entry: LookupEntry): Triple<UInt, UInt, UInt> = Triple(
            entry and 0b111111u,
            (entry shr 6) and 0b111111u,
            (entry shr 12) and 0b111111u
        )

        /**
         * Builds a lookup table with the specified offset.
         *
         * @param offset The offset to apply during table construction
         * @return New LookupTable instance
         */
        fun build(offset: Int): LookupTable {
            val output = Array(256) { EMPTY_LOOKUP_ENTRY }

            for (skip in 0..0b11111) {
                for (aux in 0..0b111) {
                    val imm1Length = minOf(4u, aux.toUInt())
                    val imm2Length = (skip.toInt() - imm1Length.toInt() - offset)
                        .coerceIn(0..4)
                        .toUInt()

                    val imm1Bits = signExtendCutoffForLength(imm1Length)
                    val imm2Bits = signExtendCutoffForLength(imm2Length)
                    val imm1Skip = imm1Length * 8u

                    val index = getLookupIndex(skip.toUInt(), aux.toUInt())
                    output[index.toInt()] = pack(imm1Bits, imm1Skip, imm2Bits)
                }
            }

            return LookupTable(output)
        }

        /**
         * Calculates lookup index from skip and aux values.
         *
         * @param skip Skip value (must be <= 0b11111)
         * @param aux Auxiliary value (must be <= 0b111)
         * @return Calculated lookup index
         */
        private fun getLookupIndex(skip: UInt, aux: UInt): UInt {
            require(skip <= 0b11111u) { "Skip value out of range" }
            val index = skip or ((aux and 0b111u) shl 5)
            require(index <= 0xFFu) { "Calculated index out of range" }
            return index
        }

        /**
         * Determines sign extension cutoff based on length.
         *
         * @param length Length value to process
         * @return Sign extension cutoff value
         */
        private fun signExtendCutoffForLength(length: UInt): UInt = when (length) {
            0u -> 32u
            1u -> 24u
            2u -> 16u
            3u -> 8u
            4u -> 0u
            else -> throw IllegalArgumentException("Invalid length: $length")
        }
    }

    /**
     * Retrieves the unpacked values for given skip and aux parameters.
     *
     * @param skip Skip value
     * @param aux Auxiliary value
     * @return Triple of (imm1Bits, imm1Skip, imm2Bits)
     */
    fun get(skip: UInt, aux: UInt): Triple<UInt, UInt, UInt> {
        val index = getLookupIndex(skip, aux)
        return unpack(table[index.toInt()])
    }
}

File: modules/pvm/src/main/scala/io/forge/jam/pvm/program/LookupTable.scala
package io.forge.jam.pvm.program

/**
 * Lookup table for handling immediate value operations.
 *
 * Provides efficient lookup for immediate value bit patterns during instruction decoding.
 * The table is indexed by (skip, aux) pairs and returns (imm1Bits, imm1Skip, imm2Bits).
 *
 * @param table The underlying array of packed lookup entries
 */
final class LookupTable private (private val table: Array[Int]):

  /**
   * Get the unpacked values for given skip and aux parameters.
   *
   * @param skip Skip value (0-31)
   * @param aux Auxiliary value
   * @return Triple of (imm1Bits, imm1Skip, imm2Bits)
   */
  def get(skip: Int, aux: Long): (Int, Int, Int) =
    val index = getLookupIndex(skip, aux)
    unpack(table(index))

  private def getLookupIndex(skip: Int, aux: Long): Int =
    skip | (((aux.toInt) & 0x7) << 5)

  private def unpack(entry: Int): (Int, Int, Int) =
    val imm1Bits = entry & 0x3F
    val imm1Skip = (entry >> 6) & 0x3F
    val imm2Bits = (entry >> 12) & 0x3F
    (imm1Bits, imm1Skip, imm2Bits)

object LookupTable:

  /** Table with offset 1 (used for most argument parsing) */
  val Table1: LookupTable = build(1)

  /** Table with offset 2 (used for regs2imm2 parsing) */
  val Table2: LookupTable = build(2)

  /**
   * Build a lookup table with the specified offset.
   *
   * @param offset The offset to apply during table construction
   * @return New LookupTable instance
   */
  def build(offset: Int): LookupTable =
    val output = new Array[Int](256)

    var skip = 0
    while skip <= 0x1F do
      var aux = 0
      while aux <= 0x7 do
        val imm1Length = math.min(4, aux)
        val imm2Length = math.max(0, math.min(4, skip - imm1Length - offset))

        val imm1Bits = signExtendCutoffForLength(imm1Length)
        val imm2Bits = signExtendCutoffForLength(imm2Length)
        val imm1Skip = imm1Length * 8

        val index = skip | (aux << 5)
        output(index) = pack(imm1Bits, imm1Skip, imm2Bits)

        aux += 1
      skip += 1

    new LookupTable(output)

  private def pack(imm1Bits: Int, imm1Skip: Int, imm2Bits: Int): Int =
    require(imm1Bits <= 0x3F, s"imm1Bits out of range: $imm1Bits")
    require(imm1Skip <= 0x3F, s"imm1Skip out of range: $imm1Skip")
    require(imm2Bits <= 0x3F, s"imm2Bits out of range: $imm2Bits")
    imm1Bits | (imm1Skip << 6) | (imm2Bits << 12)

  private def signExtendCutoffForLength(length: Int): Int = length match
    case 0 => 32
    case 1 => 24
    case 2 => 16
    case 3 => 8
    case 4 => 0
    case _ => throw new IllegalArgumentException(s"Invalid length: $length")
